File: src/parser.rs
/*
nanami -- Nano Markdown compiler
Copyright (C) 2025 pokeghost

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published
by the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

use nom::{
    bytes::complete::{tag, take_until},
    character::complete::{char, line_ending, multispace0, not_line_ending, tab},
    multi::many0,
    sequence::{delimited, preceded, tuple},
    IResult,
};

use crate::models::{Case, Content, Document};

fn parse_title(input: &str) -> IResult<&str, &str> {
    preceded(tuple((tab, tag("title: "))), not_line_ending)(input)
}

fn parse_text_block(input: &str) -> IResult<&str, &str> {
    preceded(
        tuple((multispace0, tag("text"), multispace0)),
        delimited(
            char('{'),
            delimited(multispace0, take_until("}"), multispace0),
            char('}'),
        ),
    )(input)
}

fn parse_case_name(input: &str) -> IResult<&str, &str> {
    delimited(char('('), take_until(")"), char(')'))(input)
}

fn parse_case(input: &str) -> IResult<&str, Case> {
    let (input, _) = multispace0(input)?;
    let (input, _) = tag("case")(input)?;
    let (input, name) = parse_case_name(input)?;
    let (input, _) = multispace0(input)?;
    let (input, _) = char('{')(input)?;

    let (input, texts) = many0(parse_text_block)(input)?;

    let (input, _) = multispace0(input)?;
    let (input, _) = char('}')(input)?;

    Ok((input, Case::new(name, texts)))
}

fn parse_content(input: &str) -> IResult<&str, Content> {
    preceded(
        tuple((
            multispace0,
            tag("content"),
            multispace0,
            char('{'),
            multispace0,
        )),
        delimited(
            multispace0,
            many0(parse_case),
            tuple((multispace0, char('}'))),
        ),
    )(input)
    .map(|(next_input, cases)| (next_input, Content::new(cases)))
}

pub fn parse_document(input: &str) -> IResult<&str, Document> {
    let (input, title) = parse_title(input)?;
    let (input, _) = line_ending(input)?;
    let (input, content) = parse_content(input)?;

    Ok((input, Document::new(title, content)))
}

File: lib/nanami/parser.rb
# nanami -- Nano Markdown compiler
# Copyright (C) 2025 pokeghost
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published
# by the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

# frozen_string_literal: true

require 'parslet'

module Nanami
  class Parser < Parslet::Parser
    rule(:space) { match('\s').repeat(1) }
    rule(:space?) { space.maybe }
    rule(:newline) { match('\n') }
    rule(:lbrace) { str('{') }
    rule(:rbrace) { str('}') }

    rule(:title) do
      str('title:') >> space? >> (newline.absent? >> any).repeat(1).as(:title) >> newline
    end

    # ${ref}
    rule(:ref) do
      str('$') >> lbrace >>
        match['a-zA-Z0-9_'].repeat(1).as(:ref) >>
        rbrace
    end

    rule(:html_attribute_value) { str('"') >> (str('"').absent? >> any).repeat >> str('"') }

    rule(:html_attribute) { match['a-zA-Z'].repeat(1) >> (str('=') >> html_attribute_value).maybe }

    rule(:html_tag) do
      str('<') >>
        str('/').maybe >>
        match['a-zA-Z'].repeat(1).as(:tag_name) >>
        (space >> html_attribute).repeat.as(:attributes) >>
        space? >>
        str('>')
    end

    rule(:self_closing_tag) do
      str('<') >>
        match['a-zA-Z'].repeat(1).as(:tag_name) >>
        (space >> html_attribute).repeat.as(:attributes) >>
        space? >>
        str('/>')
    end

    rule(:url_char) do
      match['a-zA-Z0-9\-._~:/?#\[\]@!$&\'*+,;=']
    end

    rule(:path_char) do
      match['a-zA-Z0-9\-._/']
    end

    # {url}{text}
    rule(:link) do
      lbrace >>
        url_char.repeat(1).as(:url) >>
        rbrace >> lbrace >>
        (lbrace.absent? >> rbrace.absent? >> any).repeat(1).as(:link_text) >>
        rbrace
    end

    # {path.ff}{description}
    rule(:image) do
      lbrace >>
        path_char.repeat(1).as(:img_path) >>
        rbrace >> lbrace >>
        (lbrace.absent? >> rbrace.absent? >> any).repeat(1).as(:img_desc) >>
        rbrace
    end

    rule(:text_content) do
      (
        ref.as(:ref) |
          image.as(:img) |
          link.as(:link) |
          self_closing_tag.as(:self_closing) |
          html_tag.as(:html) |
          (ref.absent? >> image.absent? >> link.absent? >>
            html_tag.absent? >> self_closing_tag.absent? >> str('}').absent? >> any).repeat(1).as(:plain)
      ).repeat(1)
    end

    rule(:text_block) do
      str('text') >> space? >> lbrace >> space? >>
        text_content.as(:text) >>
        space? >> rbrace
    end

    rule(:sources) do
      str('sources') >> space? >> lbrace >> space? >>
        (str('{footnotes}').as(:footnotes) >> space?).maybe >>
        rbrace
    end

    rule(:case_content) do
      (text_block | case_statement)
    end

    rule(:case_statement) do
      str('case(') >>
        (str(')').absent? >> any).repeat(1).as(:case_name) >>
        str(')') >>
        (
          str('(') >>
            url_char.repeat(1).as(:case_url) >>
            str(')')
        ).maybe >>
        space? >> lbrace >> space? >>
        case_content.repeat(0).as(:case_body) >>
        space? >> rbrace
    end

    rule(:content) do
      str('content') >> space? >> lbrace >> space? >>
        case_statement.as(:content) >> space? >>
        rbrace
    end

    rule(:document) do
      space? >> title >>
        space? >> content.as(:content)
    end

    root(:document)
  end
end
